### ISSUE TITLE
Memo: Rescue Function

### ISSUE THREAD

**Issue by aggre** [dev-protocol/DIPs#50]
_This issue is a draft DIP memo and will be used to create a DIP later, although it is not accompanied by sufficient information._

- https://discord.com/channels/547215761341546497/722710414240251984/815443284884324393
- https://twitter.com/eleon66/status/1365749611223277572

**Comment by aggre** [comment]
_Memo:_ https://discord.com/channels/547215761341546497/722710414240251984/815846692918132777

> There is currently no way to get back DEV sent directly to the contract. Technically, it is possible to add a rescue function that will only work upon the contract owner's transaction (the author of vyper). I am aware of this issue, but it is not part of the development plan yet. I have added a memo to draft a DIP at a later date. Thank you!

**Comment by Toy9cito** [comment]
Good morning I I would like to thank you for taking the time to read this I just wanted to inform you that I am one of those people that accidentally sent my Dev tokens to this address to the dev team instead of to the staking pool could you please update me on when this will be able to get my tokens back? 


![20210407_050553](https://user-images.githubusercontent.com/81981522/114388608-e9b4b800-9b61-11eb-9f75-2805333e8ce7.jpg)
![20210407_050546](https://user-images.githubusercontent.com/81981522/114388611-ea4d4e80-9b61-11eb-9736-9afccb954392.jpg)

**Comment by aggre** [comment]
Dev Protocol can solve this issue by adding the following function to `Lockup.sol`.

```solidity
function rescue(address _property, uint256 _amount, uint256 _recipient) external {
  IProperty property = IProperty(_property);
  require(msg.sender == property.author(), "not author");

  uint256 totalLocked = getStoragePropertyValue(_property);
  uint256 balance = IERC20(config().token()).balanceOf(_property);
  require(balance.sub(_amount) >= totalLocked, "no rescueable balance");
  
  property.withdraw(_recipient, _amount);
}
```

However, only the Property token author can do this. This means that the rescuee must contact the author to explain his own incorrect transfer.

@Akira-Taniguchi Do you have a better idea?

**Comment by Akira-Taniguchi** [comment]
If we leave it to the PROPERTY AUTHOR, accidents will occur, so I suggest the following procedure.

1.Ask Goxer to run the Lockup.rescue function.
　The only argument to Lockup.rescue is the Property address. It sends all differences between the DEV token balance held by the Property contract and the actual staking value to a specific wallet.
　The Lockup.rescue function can be executed by anyone.
　Of course, the assumption here is that the DEV token is executed by someone who has sent it in error.

2.Goxer will send etherscan transaction information at the time of the miscommunication to the management.

3.The operation will check it and figure out the address and the amount of tokens to be returned based on the transaction information in etherscan.

4.The operation executes the approve function, and Goxer executes the transferFrom.

Of course, this operation will be executed by the person with the role after it is made into a DAO.

**Comment by Toy9cito** [comment]
that sounds very promising you guys are just amazing and unique thank god for your great minds the world cant thank you enough

**Comment by aggre** [comment]
> Of course, this operation will be executed by the person with the role after it is made into a DAO.

@Akira-Taniguchi That operation is likely to not working once the transition to DAO is complete. Perhaps the operators need an incentive.

How about staking the rescued tokens into the Property Tokens instead of returning them to the goxer's wallet? In that case, the author also has an incentive.

@Toy9cito Do you want to stake your rescued tokens right away?

**Comment by Akira-Taniguchi** [comment]
>How about staking the rescued tokens into the Property Tokens instead of returning them to the goxer's wallet? In that case, the author also has an incentive.

This is a good idea.

　・We need to create a new lockup function to maintain internal consistency.
　・we need to create such as account_lockup table record.

I think this is Issue.

@aggre

**Comment by Toy9cito** [comment]
Ok so are going to do what exactly w my tokens..?
What are property tokens referring to the tokens that where mine sent to
the contract? Btw here is a pic of of contract transactions

**Comment by Toy9cito** [comment]
Stake them into the dev staking wout my metamask??

**Comment by Toy9cito** [comment]
How would I even be able to withdraw my tokens of they were just staked for
me wout my metamask?

**Comment by aggre** [comment]
@Toy9cito I've confused you because some explanations were missing. "Property" is a tokenized asset (OSS). Staking usually requires the `Deposit` transaction. However, some people send DEV tokens directly to a Property (by mistake). There are three proposals as a remedy for that.

1. Contact the Property author from you, and you will get back the DEV tokens by a transaction by the Property author. (DEV tokens will be back in your wallet. But the Property author has no incentive to pay for gas fees.)
2. Contact the Property author from you, and you will start staking on the Property by a transaction by the Property author. (DEV tokens are treated as if your staking was done from this operation. The Property author has an incentive to pay for gas fees.)
3. Move all fraudulently sent DEV tokens to a temporary rescue pool and return them to you by an administrator operation. (DEV tokens will be back in your wallet)

If your goal is to stake property instead of regaining DEV, plan 2 will achieve that with the fewest steps. So I asked you.

**Comment by Toy9cito** [comment]
And if I choose to have the tokens back to myself that would be more
difficult?

Also I understand your point about the contract holder ( Dev token) having
no incentive to pay for Gas fees to send the tokens back to my address..

So then the tokens will be placed in a temporary pool in which I will be
giving some kind of slot so the author (holder of the misplaced tokens)
knows which tokens are mine.. Then they will stake my tokens for me..? How
will they know me and who will initiate?

**Comment by aggre** [comment]
@Toy9cito It will be a lot difficult to get it back by yourself, and its implementation will not be realistic. Because smart contracts do not know about the DEV tokens you have transferred in the past (although I can found the transaction that you sent DEV in Etherscan). But Plan 3 is the plan that most closely matches your needs.

@Akira-Taniguchi With [Merkle Distributor](https://github.com/Uniswap/merkle-distributor), self-claiming rescue for a large number of goxers may be possible. (But it requires to set up by core maintainers in advance) On the other hand, implementing with Merkle Distributor also requires building a UI for claiming, so it might be a good idea to start with temporary support from the core maintainers. In short, Plan 3.

> Move all fraudulently sent DEV tokens to a temporary rescue pool and return them to you by an administrator operation. (DEV tokens will be back in your wallet)

**Comment by Toy9cito** [comment]
What do I need to do? "

what is suggested that I can do next?

how can I help you?

Placing that address and it saying stake just confuses people easily
Show quoted text

**Comment by Akira-Taniguchi** [comment]
@aggre 

Yes, I think proposal 3 is a good idea.
And let's make it in a way that the delegated operator will handle it when it is DAO.

**Comment by Akira-Taniguchi** [comment]
Technically it is easy, but what about the cost of gas?
For example, the operator should not have to pay for a large amount of gas to rescue 1DEV.

And the cost of the work should be received in advance.

Here is a suggestion

1.Have the operator pay 1ether in advance.
　The breakdown is the cost of gas plus the cost of work.

2.Collect the gas fee to the property->pool contract at market price.
　The pool contract should have a function that allows users to withdraw DEV tokens. When the user withdraws the DEV token, he pays for the gas himself.

Personally, I think it's idea 1.

@aggre

**Comment by Toy9cito** [comment]
Can I still communicate w you thru here until after I confirm that I have
received my tokens?

**Comment by Toy9cito** [comment]
Ok who is this operator u are saying I need to speak and communicate with?
You are my contact and I appreciate it very much..

How and with whom do u communicate with "Operator"

**Comment by Akira-Taniguchi** [comment]
Please wait a little longer as we are discussing the flow.
The operator will probably be either @aggre or @Akira-Taniguchi.

@Toy9cito

**Comment by vaanguard1** [comment]
Hey everyone, it’s great hearing discussion on this matter. 

**0x5caf454ba92e6f2c929df14667ee360ed9fd5b26**

^ I staked my dev tokens on stakes.social to this address (Vyperlang) 

Would like to ask if there is any updates on my case yet? And should I provide the full info or link from etherscan.io? (If additional verification of my statement is required)

thank you kindly,
-eleon66

> _This issue is a draft DIP memo and will be used to create a DIP later, although it is not accompanied by sufficient information._
> 
> * https://discord.com/channels/547215761341546497/722710414240251984/815443284884324393
> * https://twitter.com/eleon66/status/1365749611223277572

**Comment by Akira-Taniguchi** [comment]
Thank you for contacting me.
I'm still in the process of designing the flow, so I'll contact you via Discord or something when everything is done.

@vaanguard1

**Comment by Toy9cito** [comment]
thanks for also joining us hwere to try to help all of us here that unfortunately we sent out tokens to the contract instead of to stake our favorite project..these guys have been bery helpfull with trying to come up for a solution to get our coins back and pay for gas to have them sent back to our account,the great guys that have been helping me are The operator will probably be either @aggre or @Akira-Taniguchi.

**Comment by vaanguard1** [comment]
Yea man glad that it’s being looked into =)

thanks all

**Comment by Toy9cito** [comment]
Any news or updates on this?

**Comment by Toy9cito** [comment]
> Yea man glad that it’s being looked into =)
> 
> thanks all

good evening here and hope you having a good weekend.. just wanted to touch base and ask if you have you heard any news or updates  with the implementation of the token returns??

**Comment by timeisprecious123** [comment]
good morning...i am so happy i was directed to this thread as you know i also sent my dev tokens to the address on the page instead of thru metamask i am happy to see i wasnt the only one w that mistake is there any time line on when this issue might be addresssed ? thanks in advanced for your help 
take care and have a great day

**Comment by Toy9cito** [comment]
> Please wait a little longer as we are discussing the flow.
> The operator will probably be either @aggre or @Akira-Taniguchi.
> 
> @Toy9cito

Good day i just wanted to communicate w you and ask if we are still on schedule any closer to a solution thanks and have a gfeat day ..

i thank you so much  for all your help

**Comment by Toy9cito** [comment]
Excuse me sir sorry to bother you again but it's been quite a few months
now and I was wondering it can you tell me how of data is on returning
those coins if I pay for gas they were sent to the contract address thank
you and I hope to hear from you soon those coins are a lot of my savings
Thanks for any help you can give me
